File: Applications/ZylkerKart_Ecommerce/services/auth-service/chaos-sdk/HeartbeatWriter.cs
// Site24x7 Labs Chaos SDK for .NET — Heartbeat Writer
//
// Writes {app_name}.heartbeat.json every 30 seconds so the agent can
// discover that this SDK is installed and running.
// Uses atomic write (tmp + File.Move overwrite) to prevent partial reads.

using System.Text.Json;

namespace Site24x7.Chaos;

/// <summary>
/// Periodically writes a heartbeat JSON file so the agent can detect
/// that this SDK is installed and active. Runs as a background task
/// using <see cref="PeriodicTimer"/>.
/// </summary>
public sealed class HeartbeatWriter : IDisposable
{
    private static readonly TimeSpan HeartbeatInterval = TimeSpan.FromSeconds(30);

    private readonly string _heartbeatPath;
    private readonly string _tmpPath;
    private readonly string _appName;
    private readonly string _framework;
    private readonly string _frameworkVersion;

    private CancellationTokenSource? _cts;
    private Task? _writeTask;

    /// <summary>
    /// Creates a heartbeat writer.
    /// </summary>
    /// <param name="configDir">Directory where heartbeat files are written.</param>
    /// <param name="appName">Application name.</param>
    /// <param name="framework">Framework name (e.g., "aspnetcore").</param>
    /// <param name="frameworkVersion">Framework version string.</param>
    public HeartbeatWriter(
        string configDir,
        string appName,
        string framework,
        string frameworkVersion)
    {
        _heartbeatPath = Path.Combine(configDir, $"{appName}.heartbeat.json");
        _tmpPath = Path.Combine(configDir, $"{appName}.heartbeat.tmp");
        _appName = appName;
        _framework = framework;
        _frameworkVersion = frameworkVersion;
    }

    /// <summary>Path to the heartbeat file.</summary>
    public string HeartbeatPath => _heartbeatPath;

    // ------------------------------------------------------------------
    // Lifecycle
    // ------------------------------------------------------------------

    /// <summary>
    /// Start writing heartbeats in a background task.
    /// </summary>
    public void Start()
    {
        if (_cts is not null) return;

        // Ensure directory exists
        try
        {
            var dir = Path.GetDirectoryName(_heartbeatPath);
            if (dir is not null)
                Directory.CreateDirectory(dir);
        }
        catch
        {
            // Directory may not be writable — we'll log on first write failure
        }

        _cts = new CancellationTokenSource();
        var token = _cts.Token;

        // Write immediately
        WriteHeartbeat();

        _writeTask = Task.Run(async () =>
        {
            using var timer = new PeriodicTimer(HeartbeatInterval);
            try
            {
                while (await timer.WaitForNextTickAsync(token))
                {
                    WriteHeartbeat();
                }
            }
            catch (OperationCanceledException)
            {
                // Expected on shutdown
            }
        }, token);

        Console.WriteLine($"[chaos-sdk] HeartbeatWriter started, path: {_heartbeatPath}");
    }

    /// <summary>
    /// Stop the heartbeat writer, wait for the background task, and remove the heartbeat file.
    /// </summary>
    public void Stop()
    {
        if (_cts is null) return;

        _cts.Cancel();

        try { _writeTask?.Wait(); }
        catch (AggregateException) { /* cancelled */ }

        _cts.Dispose();
        _cts = null;
        _writeTask = null;

        // Clean up heartbeat file on shutdown
        try { File.Delete(_heartbeatPath); }
        catch { /* may not exist */ }

        Console.WriteLine("[chaos-sdk] HeartbeatWriter stopped");
    }

    // ------------------------------------------------------------------
    // Internal
    // ------------------------------------------------------------------

    private void WriteHeartbeat()
    {
        var data = new Dictionary<string, object>
        {
            ["app_name"] = _appName,
            ["sdk_version"] = ChaosConstants.SdkVersion,
            ["sdk_language"] = ChaosConstants.SdkLanguage,
            ["framework"] = _framework,
            ["framework_version"] = _frameworkVersion,
            ["pid"] = Environment.ProcessId,
            ["hostname"] = Environment.MachineName,
            ["timestamp"] = DateTime.UtcNow.ToString("yyyy-MM-ddTHH:mm:ssZ"),
        };

        try
        {
            var json = JsonSerializer.Serialize(data, new JsonSerializerOptions { WriteIndented = true });

            // Atomic write: write to tmp file, then rename
            File.WriteAllText(_tmpPath, json);
            File.Move(_tmpPath, _heartbeatPath, overwrite: true);
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"[chaos-sdk] Failed to write heartbeat file: {ex.Message}");
        }
    }

    // ------------------------------------------------------------------
    // IDisposable
    // ------------------------------------------------------------------

    public void Dispose()
    {
        Stop();
    }
}

File: Applications/ZylkerKart_Ecommerce/services/product-service/chaos-sdk/chaos-sdk-core/src/main/java/com/site24x7/labs/chaos/config/HeartbeatWriter.java
package com.site24x7.labs.chaos.config;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.net.InetAddress;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardCopyOption;
import java.time.Instant;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Periodically writes a {@code {appName}.heartbeat.json} file so the agent
 * can detect that this SDK is installed and running.
 *
 * <p>The heartbeat file contains:
 * <ul>
 *   <li>{@code app_name} – matches the config filename stem</li>
 *   <li>{@code sdk_version}</li>
 *   <li>{@code sdk_language} – always {@code "java"}</li>
 *   <li>{@code framework} – e.g. {@code "spring-boot"}</li>
 *   <li>{@code framework_version}</li>
 *   <li>{@code pid}</li>
 *   <li>{@code hostname}</li>
 *   <li>{@code timestamp} – ISO-8601 UTC</li>
 * </ul>
 */
public class HeartbeatWriter implements Runnable {

    private static final Logger log = LoggerFactory.getLogger(HeartbeatWriter.class);
    private static final Gson GSON = new GsonBuilder().setPrettyPrinting().create();
    private static final long DEFAULT_INTERVAL_MS = 30_000;

    private final Path heartbeatPath;
    private final String appName;
    private final String sdkVersion;
    private final String framework;
    private final String frameworkVersion;
    private final long intervalMs;
    private volatile boolean running = true;

    public HeartbeatWriter(String configDir, String appName, String sdkVersion,
                           String framework, String frameworkVersion) {
        this(configDir, appName, sdkVersion, framework, frameworkVersion, DEFAULT_INTERVAL_MS);
    }

    public HeartbeatWriter(String configDir, String appName, String sdkVersion,
                           String framework, String frameworkVersion, long intervalMs) {
        this.heartbeatPath = Paths.get(configDir, appName + ".heartbeat.json");
        this.appName = appName;
        this.sdkVersion = sdkVersion;
        this.framework = framework;
        this.frameworkVersion = frameworkVersion;
        this.intervalMs = intervalMs;
    }

    @Override
    public void run() {
        log.info("HeartbeatWriter started, path: {}", heartbeatPath);
        while (running) {
            try {
                writeHeartbeat();
                Thread.sleep(intervalMs);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                break;
            } catch (Exception e) {
                log.warn("Error writing heartbeat file: {}", e.getMessage());
                try {
                    Thread.sleep(intervalMs);
                } catch (InterruptedException ie) {
                    Thread.currentThread().interrupt();
                    break;
                }
            }
        }
        // Clean up heartbeat file on shutdown
        try {
            Files.deleteIfExists(heartbeatPath);
        } catch (IOException ignored) {
        }
        log.info("HeartbeatWriter stopped");
    }

    private void writeHeartbeat() throws IOException {
        Map<String, Object> data = new LinkedHashMap<>();
        data.put("app_name", appName);
        data.put("sdk_version", sdkVersion);
        data.put("sdk_language", "java");
        data.put("framework", framework);
        data.put("framework_version", frameworkVersion);
        data.put("pid", ProcessHandle.current().pid());
        data.put("hostname", getHostname());
        data.put("timestamp", DateTimeFormatter.ISO_INSTANT.format(Instant.now().atOffset(ZoneOffset.UTC)));

        String json = GSON.toJson(data);

        // Atomic write: write to tmp then rename
        Path tmpPath = heartbeatPath.resolveSibling(heartbeatPath.getFileName() + ".tmp");
        Files.writeString(tmpPath, json);
        Files.move(tmpPath, heartbeatPath, StandardCopyOption.REPLACE_EXISTING, StandardCopyOption.ATOMIC_MOVE);
    }

    private static String getHostname() {
        try {
            return InetAddress.getLocalHost().getHostName();
        } catch (Exception e) {
            return "unknown";
        }
    }

    public void stop() {
        running = false;
    }
}
